### PR TITLE
Update to use DataPostprocessorInputs

### PR DESCRIPTION
**Pull request by class4kayaker**
In progress solution to #1268 and #1465. Confirmed to function for one location of use without additional modification, working on others.

### CODE REVIEW

**Comment by gassmoeller** [comment]
Great thanks. This looks about right. Lets see if it works for all postprocessors.
@bangerth: It seems you do not longer need to do this yourself.

**Comment by bangerth** [comment]
Ha, it seems like the old saying is still true: If you just wait for long enough, then someone else will clean up the mess you created :-)

Thanks, @class4kayaker !

**Comment by bangerth** [comment]
@class4kayaker -- I agree with @gassmoeller that this looks right. Do you think you will be able to finish this sometime soon?

**Comment by class4kayaker** [comment]
I have just finished all but the melt visualization postprocessor, which I had been running into wierd issues with earlier, but did just compile. I am running a final local test before uploading the new squashed commit.

**Comment by class4kayaker** [comment]
Revised all functions of compute_derived_quantities_vector to use evaluate_vector_field, should be ready to merge after review.

File: include/aspect/postprocess/visualization/adiabat.h [hunk 1]
```diff
@@ -59,12 +59,8 @@ namespace aspect
...
           virtual
           void
-          compute_derived_quantities_vector (const std::vector<Vector<double> >              &solution_values,
-                                             const std::vector<std::vector<Tensor<1,dim> > > &solution_gradients,
-                                             const std::vector<std::vector<Tensor<2,dim> > > &solution_hessians,
-                                             const std::vector<Point<dim> >                  &normals,
-                                             const std::vector<Point<dim> >                  &evaluation_points,
-                                             std::vector<Vector<double> >                    &computed_quantities) const;
+          evaluate_vector_field(const DataPostprocessorInputs::Vector<dim> &input_data,
+                                std::vector<Vector<double>> &computed_quantities) const;
```
Comment (gassmoeller) [inline include/aspect/postprocess/visualization/adiabat.h]: `>>` will be interpreted by older compilers as the operator `>>`. Please change it to `> >`.
Same in all other places.

File: source/postprocess/visualization/viscosity.cc [hunk 2]
```diff
@@ -42,42 +42,38 @@ namespace aspect
...
 
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
+            in.temperature[q]=input_data.solution_values[q][this->introspection().component_indices.temperature];
             for (unsigned int d = 0; d < dim; ++d)
               {
-                in.velocity[q][d]=solution_values[q][this->introspection().component_indices.velocities[d]];
-                in.pressure_gradient[q][d] = solution_gradients[q][this->introspection().component_indices.pressure][d];
+                in.velocity[q][d]=input_data.solution_values[q][this->introspection().component_indices.velocities[d]];
```
Comment (gassmoeller) [inline source/postprocess/visualization/viscosity.cc]: missing space

File: source/postprocess/visualization/viscosity.cc [hunk 3]
```diff
@@ -42,42 +42,38 @@ namespace aspect
...
             Tensor<2,dim> grad_u;
             for (unsigned int d=0; d<dim; ++d)
-              grad_u[d] = solution_gradients[q][d];
+              grad_u[d] = input_data.solution_gradients[q][d];
             in.strain_rate[q] = symmetrize (grad_u);
 
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
+            in.temperature[q]=input_data.solution_values[q][this->introspection().component_indices.temperature];
```
Comment (gassmoeller) [inline source/postprocess/visualization/viscosity.cc]: missing spaces

File: source/postprocess/visualization/vertical_heat_flux.cc [hunk 4]
```diff
@@ -63,21 +59,21 @@ namespace aspect
...
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
+            in.temperature[q]=input_data.solution_values[q][this->introspection().component_indices.temperature];
             for (unsigned int d = 0; d < dim; ++d)
               {
-                in.velocity[q][d]=solution_values[q][this->introspection().component_indices.velocities[d]];
-                in.pressure_gradient[q][d] = solution_gradients[q][this->introspection().component_indices.pressure][d];
-                temperature_gradient[q][d] = solution_gradients[q][this->introspection().component_indices.temperature][d];
+                in.velocity[q][d]=input_data.solution_values[q][this->introspection().component_indices.velocities[d]];
```
Comment (gassmoeller) [inline source/postprocess/visualization/vertical_heat_flux.cc]: space

File: source/postprocess/visualization/vertical_heat_flux.cc [hunk 5]
```diff
@@ -63,21 +59,21 @@ namespace aspect
...
 
-        in.position = evaluation_points;
+        in.position = input_data.evaluation_points;
         in.strain_rate.resize(0); // we do not need the viscosity
         for (unsigned int q=0; q<n_quadrature_points; ++q)
           {
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
+            in.temperature[q]=input_data.solution_values[q][this->introspection().component_indices.temperature];
```
Comment (gassmoeller) [inline source/postprocess/visualization/vertical_heat_flux.cc]: spaces

File: source/postprocess/visualization/thermal_expansivity.cc [hunk 6]
```diff
@@ -42,38 +42,34 @@ namespace aspect
...
             //in.strain_rate[q] =
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
+            in.temperature[q]=input_data.solution_values[q][this->introspection().component_indices.temperature];
             for (unsigned int d = 0; d < dim; ++d)
               {
-                in.velocity[q][d]=solution_values[q][this->introspection().component_indices.velocities[d]];
-                in.pressure_gradient[q][d] = solution_gradients[q][this->introspection().component_indices.pressure][d];
+                in.velocity[q][d]=input_data.solution_values[q][this->introspection().component_indices.velocities[d]];
```
Comment (gassmoeller) [inline source/postprocess/visualization/thermal_expansivity.cc]: space

File: source/postprocess/visualization/thermal_expansivity.cc [hunk 7]
```diff
@@ -42,38 +42,34 @@ namespace aspect
...
-        in.position = evaluation_points;
+        in.position = input_data.evaluation_points;
         in.strain_rate.resize(0); // we do not need the viscosity
         for (unsigned int q=0; q<n_quadrature_points; ++q)
           {
             //in.strain_rate[q] =
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
+            in.temperature[q]=input_data.solution_values[q][this->introspection().component_indices.temperature];
```
Comment (gassmoeller) [inline source/postprocess/visualization/thermal_expansivity.cc]: spaces

File: source/postprocess/visualization/thermal_diffusivity.cc [hunk 8]
```diff
@@ -42,32 +42,28 @@ namespace aspect
...
                                                      this->n_compositional_fields());
 
-        in.position = evaluation_points;
+        in.position = input_data.evaluation_points;
         in.strain_rate.resize(0); // we do not need the viscosity
         for (unsigned int q=0; q<n_quadrature_points; ++q)
           {
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
```
Comment (gassmoeller) [inline source/postprocess/visualization/thermal_diffusivity.cc]: space

File: source/postprocess/visualization/thermal_conductivity.cc [hunk 9]
```diff
@@ -42,32 +42,28 @@ namespace aspect
...
                                                      this->n_compositional_fields());
 
-        in.position = evaluation_points;
+        in.position = input_data.evaluation_points;
         in.strain_rate.resize(0); // we do not need the viscosity
         for (unsigned int q=0; q<n_quadrature_points; ++q)
           {
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
```
Comment (gassmoeller) [inline source/postprocess/visualization/thermal_conductivity.cc]: space

File: source/postprocess/visualization/stress.cc [hunk 10]
```diff
@@ -32,45 +32,41 @@ namespace aspect
...
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
+            in.temperature[q]=input_data.solution_values[q][this->introspection().component_indices.temperature];
 
             for (unsigned int d = 0; d < dim; ++d)
               {
-                in.velocity[q][d]=solution_values[q][this->introspection().component_indices.velocities[d]];
-                in.pressure_gradient[q][d] = solution_gradients[q][this->introspection().component_indices.pressure][d];
+                in.velocity[q][d]=input_data.solution_values[q][this->introspection().component_indices.velocities[d]];
```
Comment (gassmoeller) [inline source/postprocess/visualization/stress.cc]: space

File: source/postprocess/visualization/stress.cc [hunk 11]
```diff
@@ -32,45 +32,41 @@ namespace aspect
...
             Tensor<2,dim> grad_u;
             for (unsigned int d=0; d<dim; ++d)
-              grad_u[d] = solution_gradients[q][d];
+              grad_u[d] = input_data.solution_gradients[q][d];
             in.strain_rate[q] = symmetrize (grad_u);
 
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
+            in.temperature[q]=input_data.solution_values[q][this->introspection().component_indices.temperature];
```
Comment (gassmoeller) [inline source/postprocess/visualization/stress.cc]: spaces

File: source/postprocess/visualization/specific_heat.cc [hunk 12]
```diff
@@ -42,37 +42,33 @@ namespace aspect
...
           {
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
+            in.temperature[q]=input_data.solution_values[q][this->introspection().component_indices.temperature];
             for (unsigned int d = 0; d < dim; ++d)
               {
-                in.velocity[q][d]=solution_values[q][this->introspection().component_indices.velocities[d]];
-                in.pressure_gradient[q][d] = solution_gradients[q][this->introspection().component_indices.pressure][d];
+                in.velocity[q][d]=input_data.solution_values[q][this->introspection().component_indices.velocities[d]];
```
Comment (gassmoeller) [inline source/postprocess/visualization/specific_heat.cc]: space

File: source/postprocess/visualization/specific_heat.cc [hunk 13]
```diff
@@ -42,37 +42,33 @@ namespace aspect
...
 
-        in.position = evaluation_points;
+        in.position = input_data.evaluation_points;
         in.strain_rate.resize(0); // we do not need the viscosity
         for (unsigned int q=0; q<n_quadrature_points; ++q)
           {
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
+            in.temperature[q]=input_data.solution_values[q][this->introspection().component_indices.temperature];
```
Comment (gassmoeller) [inline source/postprocess/visualization/specific_heat.cc]: spaces

File: source/postprocess/visualization/shear_stress.cc [hunk 14]
```diff
@@ -32,44 +32,40 @@ namespace aspect
...
 
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
+            in.temperature[q]=input_data.solution_values[q][this->introspection().component_indices.temperature];
             for (unsigned int d = 0; d < dim; ++d)
               {
-                in.velocity[q][d]=solution_values[q][this->introspection().component_indices.velocities[d]];
-                in.pressure_gradient[q][d] = solution_gradients[q][this->introspection().component_indices.pressure][d];
+                in.velocity[q][d]=input_data.solution_values[q][this->introspection().component_indices.velocities[d]];
```
Comment (gassmoeller) [inline source/postprocess/visualization/shear_stress.cc]: space

File: source/postprocess/visualization/shear_stress.cc [hunk 15]
```diff
@@ -32,44 +32,40 @@ namespace aspect
...
             Tensor<2,dim> grad_u;
             for (unsigned int d=0; d<dim; ++d)
-              grad_u[d] = solution_gradients[q][d];
+              grad_u[d] = input_data.solution_gradients[q][d];
             in.strain_rate[q] = symmetrize (grad_u);
 
-            in.pressure[q]=solution_values[q][this->introspection().component_indices.pressure];
-            in.temperature[q]=solution_values[q][this->introspection().component_indices.temperature];
+            in.pressure[q]=input_data.solution_values[q][this->introspection().component_indices.pressure];
+            in.temperature[q]=input_data.solution_values[q][this->introspection().component_indices.temperature];
```
Comment (gassmoeller) [inline source/postprocess/visualization/shear_stress.cc]: spaces

File: source/postprocess/visualization/shear_stress.cc [hunk 16]
```diff
@@ -32,44 +32,40 @@ namespace aspect
...
       void
       ShearStress<dim>::
-      compute_derived_quantities_vector (const std::vector<Vector<double> >              &solution_values,
-                                         const std::vector<std::vector<Tensor<1,dim> > > &solution_gradients,
-                                         const std::vector<std::vector<Tensor<2,dim> > > &,
-                                         const std::vector<Point<dim> > &,
-                                         const std::vector<Point<dim> >                  &evaluation_points,
-                                         std::vector<Vector<double> >                    &computed_quantities) const
+      evaluate_vector_field(const DataPostprocessorInputs::Vector<dim> &input_data,
+                            std::vector<Vector<double>> &computed_quantities) const
```
Comment (gassmoeller) [inline source/postprocess/visualization/shear_stress.cc]: same issue here as in the header files, please replace all `>>` by `> >`

**Comment by class4kayaker** [comment]
Fixed the cases with`>>` to `> >`.

For adding the spaces, are you referring to adding spaces around the `=` operator, because in many of these lines, there were none originally, so what you are referring to is somewhat ambiguous.

**Comment by gassmoeller** [comment]
Ah right. Our usual convection is to have a space before an assignment equal sign, and one behind it. I did not see that there were wrong lines before. Still it would be nice to fix them, when we are touching them anyway.

**Comment by class4kayaker** [comment]
Makes sense, I just wanted to confirm that that was the desired modification.

**Comment by class4kayaker** [comment]
I have added spaces around the `=` operator in the visualization post-processors touched in this pull request.

**Comment by gassmoeller** [comment]
Looks good. Thanks! Ready to merge if the tester is happy.

**Pull request merged**